Argo/Globals/Parser.swift

```diff
@@ -0,0 +1,26 @@
+import Box
+
+public enum Parser<T> {
+  case Success(Box<T>)
+  case TypeMismatch(String)
+  case MissingKey(String)
+
+  public var value: T? {
+    switch self {
+    case let .Success(box): return box.value
+    default: return .None
+    }
+  }
+
+  public static func typeMismatch(expectedType: String, object: JSON) -> Parser<T> {
+    return .TypeMismatch("\(object) is not a \(expectedType)")
+  }
+}
+
+func parseOptional<A>(p: Parser<A>) -> Parser<A?> {
```

Comment:
Bringing back the conversation about this sucker:

Is it possible that this is doing the wrong thing, and that's why it's hard to name? Is it possible that `.TypeMismatch` should _also_ be coerced into `.Success(.None)`? If that were the case, it seems like it would simplify this function conceptually, and also seems like it might end up with a clearer name?